File: mqttManagerLib/MQTTManager.cs
using MqttLib;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MQTTManagerLib
{
    /// <summary>
    /// Based on MQTT Server : https://www.cloudmqtt.com/docs-dotnet.html
    /// Mqtt lib : https://github.com/stevenlovegrove/MqttDotNet
    /// </summary>
    public class MQTTManager : IDisposable
    {
        private string _connectionString;
        private string _clientId;
        private IMqtt _client;

        public delegate void NotificationEventType(string message);
        public delegate void MessageArrivedType(MQTTMessage message);
        public event NotificationEventType NotificationEvent;
        public event MessageArrivedType MessageArrived;

        private void Notify(string m)
        {
            if(NotificationEvent != null)
                NotificationEvent(m);
        }

        public MQTTManager(string connectionString, string clientId, string userName, string password)
        {
            this._connectionString = connectionString;
            this._clientId = clientId;
            this._client = MqttClientFactory.CreateClient(connectionString, _clientId, userName, password);
            this._client.Connected += new ConnectionDelegate(client_Connected);
            this._client.ConnectionLost += new ConnectionDelegate(_client_ConnectionLost);
            this._client.PublishArrived += new PublishArrivedDelegate(client_PublishArrived);
        }

        public static string BuildClientId()
        {
            var g = Guid.NewGuid().ToString().Split('-');
            return $"{Environment.MachineName}-{g[0]}";
        }

        public void Start(string channel)
        {
            Notify("Client connecting");
            this._client.Connect(true);
            this.Publish(channel, $"New console instance connected");
        }

        public void Stop()
        {
            if (this._client.IsConnected)
            {
                this.Notify("Client disconnecting");
                this._client.Disconnect();
                this.Notify("Client disconnected");
            }
        }
        public void Subscribe(string channel)
        {
            var r = this._client.Subscribe(channel, QoS.BestEfforts);
            this.Notify($"Client Subscribed '{channel}'");
        }
        private string BuildMessage(string message, string topic)
        {
            var m = new MQTTMessage() {
                Message = message,
                ClientId = this._clientId,
                Topic = topic
            };
            return m.ToJSON();
        }
        public void Publish(string channel, string message)
        {
            this._client.Publish(channel, BuildMessage(message, channel), QoS.BestEfforts, false);
        }
        void client_Connected(object sender, EventArgs e)
        {
            this.Notify("Client connected");
        }
        void _client_ConnectionLost(object sender, EventArgs e)
        {
            this.Notify("Client connection lost");
        }
        bool client_PublishArrived(object sender, PublishArrivedArgs e)
        {
            var m = MQTTMessage.Parse(e.Payload.ToString(), e.Topic);
            if(m.ClientId == this._clientId)
            {
                // Ignore message sent by this instance
                return false;
            }
            else {
                if(this.MessageArrived != null)
                    this.MessageArrived(m);
                return true;
            }
        }
        public void Dispose()
        {
            this.Stop();
        }
    }
}

File: mqttManagerLib/JavaScript/mqttManager.js
/*
    mqttManager.js a wrapper/helper around MQTT.js for browser.
    - Developed to work with MQTT broker from CloudMqtt.com.
    - Tested in 
        - Chrome Windows and Chrome iOS.
        - Edge and IE 11 Windows.

    Dependencies:
    - mqtt.js
        https://github.com/mqttjs/MQTT.js
        c:\>npm install mqtt --save

    Frederic Torres 2018.10
 */
var mqtt = require('mqtt');

const WEB_SOCKET_DEFAULT_PORT = 30989;

class MqttManager {

    _client = null;
    _options = null;

    constructor(mqttUrl, username, password, webSocketPort = WEB_SOCKET_DEFAULT_PORT, clientId) {

        if (!clientId)
            clientId = 'mqttManager.js_' + Math.random().toString(16).substr(2, 8);

        this._options = {
            port: webSocketPort, // Required for WebSocket
            useSSL: true,        // Required for WebSocket
            host: mqttUrl,
            clientId,
            username,
            password,
            keepalive: 60,
            reconnectPeriod: 1000,
            protocolId: 'MQIsdp',
            protocolVersion: 3,
            clean: true,
            encoding: 'utf8'
        };
    }
    trace(m) {
        console.log(`[MqttMgr]${m}`);
    }
    parseMessage(rawMessage, topic) {
        const o = JSON.parse(rawMessage);
        const oo = { ...o, topic };
        console.log('PARSE MESSAGE', JSON.stringify(oo));
        return oo;
    }
    onReceivedMessage = (topic, message) => {
        const parsedMessage = this.parseMessage(message, topic);
        if (parsedMessage.clientId === this._options.clientId) {
            return; // ignore message sent by this instance
        }
        this.trace(`RECEIVED:${parsedMessage.message}, ${parsedMessage.clientId}, ${parsedMessage.topic}`);
        if (this.messageArrived) {
            this.messageArrived(parsedMessage);
        }
    }
    subscribe(channel) {
        return new Promise((resolve, reject) => {
            this.trace(`Subscribe ${channel}`);
            this._client.subscribe(channel, (err) => {
                if (err) {
                    console.log(`Subscribe err:${err}`);
                    reject(channel);
                }
                else {
                    this.trace(`Subscribed ${channel}`);
                    this._client.on('message', this.onReceivedMessage);
                    resolve(channel);
                }
            });
        });
    }
    buildMessage(message) {
        const s = JSON.stringify({ message, clientId: this._options.clientId });
        console.log('BUILD MESSAGE', s);
        return s;
    }
    publish(channel, message) {
        return new Promise((resolve, reject) => {
            this.trace(`Publishing message:${message}`);
            // https://github.com/mqttjs/MQTT.js#publish
            this._client.publish(channel, this.buildMessage(message), undefined, (err) => {
                if (err) {
                    this.trace(`Publishing failed message:${message}`);
                    reject(message);
                }
                else {
                    resolve(message);
                }
            });
        });
    }
    start() {
        return new Promise((resolve, reject) => {

            const optionsOkToTrace = Object.assign({ password: 'ok' }, this._options);
            optionsOkToTrace.password = null;
            this.trace(`Connecting options:${JSON.stringify(optionsOkToTrace)}`);

            this._client = mqtt.connect(this._options.host, this._options);

            this._client.on('error', (err) => {
                console.log(err);
                reject(false);
            });
            this._client.on('connect', () => {
                this.trace(`MQTT Connected to ${this._options.host}`);
                resolve(true);
            });
        });
    }
    stop() {

    }
}

export default MqttManager;
